monitoring/opencensus/metrics.go

```diff
@@ -0,0 +1,327 @@
+package opencensus
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math"
+	"strings"
+	"time"
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	"github.com/golang/glog"
+	"github.com/google/trillian/monitoring"
+
+	"go.opencensus.io/exporter/prometheus"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+
+	datadog "github.com/Datadog/opencensus-go-exporter-datadog"
+)
+
+const (
+	namespace = "trillian"
+	separator = "_"
+)
+
+// MetricFactory allows the creation of OpenCensus measures and views
+type MetricFactory struct {
+	Prefix string
+}
+
+var (
+// Stackdriver Client will be used by [Custom|Gauge|Histogram] types to compute Value(s)
+//client *stackdriver_client.MetricClient
+)
+
+// init initializes Stackdriver, Prometheus and Datadog exporters
+// once registered with Opencensus' View, all metrics will export to these systems
+// Only Stackdriver requires a 60-second reporting period but this met be set once for all
+// Stackdriver authentication uses Application Default Credentials and assumes
+// GOOGLE_APPLICATION_CREDENTIALS references a service account key
+// The service account permissions must include at least roles/monitoring.metricWriter
+// Prometheus should expose a /metrics endpoint but currently does not
+// Datadog exporter assumes the Datadog Agent is running on localhost
+//TODO(dazwilkin) Is this the best place to initialize the exporters?
+//TODO(dazwilkin) OpenCensus exporters should be configured by config
+func init() {
+	// Stackdriver Exporter
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
+		// MetricPrefix helps uniquely identify these metrics
+		//TODO(dazwilkin) How to create the exporter in order to use MetricFactory.Prefix
+		MetricPrefix: namespace,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Important to invoke Flush before exiting
+	defer sd.Flush()
```

Comment:
Addressed thanks to your guidance: https://github.com/google/trillian/pull/1414/commits/2706a6e3eadc500bce61a329125c062cd0457ad3

This requires changes to each server `main` which would break the Prometheus implementation. So, leaving unchanged for now. Must also understand how best to reflect the configuration through Trillian's configuration.

```golang
//TODO(dazwilkin) Datadog, Prometheus & Stackdriver should be configurable
if true {
	flush, err := opencensus.Initialize()
	if err != nil {
		glog.Exitf("Failed to initialize OpenCensus Exporter(s): %v", err)
	}
	defer flush()
}
```